lib/include/tick/array/sparsearray2d.h

```diff
@@ -107,26 +107,41 @@ class SparseArray2d : public BaseArray2d<T> {
     ar(this->_size);
 
     inner_save(ar);
+  }
 
-    std::vector<INDICE_TYPE> v_indices(this->_indices, this->_indices + this->_size_sparse),
-        v_row_indices(this->_row_indices, this->_row_indices + (this->_n_rows + 1));
-    ar(v_indices);
-    ar(v_row_indices);
+  template <class Archive, typename Y = T>
+  typename std::enable_if<std::is_same<Y, std::atomic<K>>::value
+    && cereal::traits::is_output_serializable<cereal::BinaryData<T>, Archive>::value, void>::type
+  inner_save(Archive &ar) const {
+    for (size_t i = 0; i < this->_size_sparse; i++) ar(this->_data[i].load());
+    ar(cereal::binary_data(this->_indices, sizeof(INDICE_TYPE) * this->_size_sparse));
+    ar(cereal::binary_data(this->_row_indices, sizeof(INDICE_TYPE) * (this->_n_rows + 1)));
   }
 
   template <class Archive, typename Y = T>
-  typename std::enable_if<std::is_same<Y, std::atomic<K>>::value>::type inner_save(
-      Archive &ar) const {
-    std::vector<K> k_data;
-    for (size_t i = 0; i < this->_size_sparse; i++) k_data.emplace_back(this->_data[i]);
-    ar(k_data);
+  typename std::enable_if<std::is_same<Y, std::atomic<K>>::value
+    && !cereal::traits::is_output_serializable<cereal::BinaryData<T>, Archive>::value, void>::type
+  inner_save(Archive &ar) const {
+    for (size_t i = 0; i < this->_size_sparse; i++) ar(this->_data[i].load());
+    for (size_t i = 0; i < this->_size_sparse; i++) ar(this->_indices[i]);
+    for (size_t i = 0; i < this->_n_rows + 1; i++) ar(this->_row_indices[i]);
   }
 
   template <class Archive, typename Y = T>
-  typename std::enable_if<!std::is_same<Y, std::atomic<K>>::value>::type inner_save(
-      Archive &ar) const {
-    std::vector<T> v_data(this->_data, this->_data + this->_size_sparse);
-    ar(v_data);
+  typename std::enable_if<!std::is_same<Y, std::atomic<K>>::value
```

Comment:
Wouldn't `std::enable_if<std::is_same<Y, K>` be clearer than `std::enable_if<!std::is_same<Y, std::atomic<K>>::value` ?
Anyway it might worth adding a comment before each function telling in which cases it applies (like Atomic && binary serialization or Non atomic && non binary-seriation). Because it's painful to decrypt these on Monday morning ;)

Reply:
it would and sure